tests/apps/test_graph_max_clique.py

```diff
@@ -0,0 +1,95 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""
+Unit tests for strawberryfields.apps.graph.max_clique
+"""
+# pylint: disable=no-self-use,unused-argument
+import pytest
+import networkx as nx
+import numpy as np
+
+from strawberryfields.apps.graph import max_clique
+from strawberryfields.apps.graph import resize
+
+
+def patch_random_choice(x):
+    """Dummy function for ``np.random.choice`` to make output deterministic. This dummy function
+    just returns the element of ``x`` specified by ``element``."""
+    if isinstance(x, int):  # np.random.choice can accept an int or 1D array-like input
+        return 0
+
+    return x[0]
+
+
+def patch_clique_resize(clique, graph, node_select):
+    """Dummy function for ``clique_grow`` and ``clique_swap`` to make output unchanged."""
+    return clique
+
+
+class TestLocalSearch:
+    """Tests for the function ``max_clique.local_search``"""
+
+    def test_bad_iterations(self):
+        """Test if function raises a ``ValueError`` when a non-positive number of iterations is
+        specified"""
+        with pytest.raises(ValueError, match="Number of iterations must be a positive int"):
+            max_clique.local_search(clique=[0, 1, 2, 3], graph=nx.empty_graph(5), iterations=-1)
+
+    def test_max_iterations(self, monkeypatch):
+        """Test if function stops after 5 iterations despite not being in a dead end (i.e., when
+        ``grow != swap``). This is achieved by monkeypatching the ``np.random.choice`` call in
+        ``clique_grow`` and ``clique_swap`` so that for a 5-node wheel graph starting as a [0,
+        1] node clique, the algorithm simply oscillates between [0, 1, 2] and [0, 2, 3] for each
+        iteration of growth & swap. For odd iterations, we get [0, 2, 3]."""
+
+        graph = nx.wheel_graph(5)
+        clique = [0, 1]
+
+        with monkeypatch.context() as m:
+            m.setattr(np.random, "choice", patch_random_choice)
+            result = max_clique.local_search(clique, graph, iterations=5)
+
+        assert result == [0, 2, 3]
+
+    def test_dead_end(self, monkeypatch):
+        """Test if function stops when in a dead end (i.e., ``grow == swap``) such that no
+        swapping is possible. This is achieved by monkeypatching ``clique_grow`` and
+        ``clique_swap`` so that they simply return the same clique as fed in, which should cause
+        the local search algorithm to conclude that it is already in a dead and and return the
+        same clique."""
+
+        graph = nx.complete_graph(5)
+        clique = [0, 1, 2]
+
+        with monkeypatch.context() as m:
+            m.setattr(resize, "clique_grow", patch_clique_resize)
+            m.setattr(resize, "clique_swap", patch_clique_resize)
+            result = max_clique.local_search(clique, graph)
+
+        assert result == clique
+
+    def test_expected_growth(self):

```

Comment:
This is probably the main test, which seems ok to me. @ixfoduap any ideas to improve it to a more complicated example?

Reply:
I think this is a great test. It checks all the main steps of the algorithm. We could come up with a similar but more complicated test, but I don't think it would actually test any additional features